src/Struct/FindologicProduct.php

```diff
@@ -451,21 +455,33 @@ protected function setImages(): void
             return;
         }
 
+        /** @var ProductMediaEntity $mediaEntity */
         foreach ($this->getSortedImages() as $mediaEntity) {
-            if (!$mediaEntity->getMedia() || !$mediaEntity->getMedia()->getUrl()) {
+            $media = $mediaEntity->getMedia();
+            if (!$media || !$media->getUrl()) {
                 continue;
             }
 
-            $encodedUrl = $this->getEncodedUrl($mediaEntity->getMedia()->getUrl());
-            if (!Utils::isEmpty($encodedUrl)) {
-                $this->images[] = new Image($encodedUrl);
-            }
-
-            $thumbnails = $mediaEntity->getMedia()->getThumbnails();
+            $thumbnails = $media->getThumbnails();
             if (!$thumbnails) {
+                $encodedUrl = $this->getEncodedUrl($media->getUrl());
+                if (!Utils::isEmpty($encodedUrl)) {
+                    $this->images[] = new Image($encodedUrl);
+                }
                 continue;
             }
 
+            $filteredThumbnails = $this->filterThumbnails($thumbnails);
+
+            // Use the thumbnail as the main image if available, otherwise fallback to the directly assigned image.
+            $image = $filteredThumbnails->first() ?? $media;
+            if ($image) {
+                $encodedThumbnailUrl = $this->getEncodedUrl($image->getUrl());
+                if (!Utils::isEmpty($encodedThumbnailUrl)) {
+                    $this->images[] = new Image($encodedThumbnailUrl);
+                }
```

Comment:
This exact block is duplicate to lines 467 - 470.

Reply:
No actually, I had already tried refactoring this, but a ThumbnailMediaEntity is different than a MediaEntity so the typehint would not match. But I will try it refactoring another way.

Reply:
As a hint, if there isn't a general class, you may also just use `Entity` as a return type and add `ThumbnailMediaEntity|MediaEntity` to the return type in the docblock.

Reply:
This wasn't addressed.